File: scala/phone-number/src/main/scala/phone_number.scala
class PhoneNumber(raw_number: String) {

    def number() = {
        if (digits.length == 10)
            digits
        else if (digits.length == 11 && digits.slice(0, 1) == "1")
            digits.slice(1, 11)
        else
            "0000000000"
    }

    def digits() = {
        "\\d".r.findAllIn(raw_number).reduce { _ + _ }
    }

    def areaCode() = {
        number.slice(0, 3)
    }

    override def toString() = {
        "(123) 456-7890"
    }
}

File: ecmascript/phone-number/phone-number.js
class PhoneNumber {
  constructor(number) {
    const clean = number.replace(/\D/g, '');
    const validated = clean.match(/^(?:1)?(\d{10})$/);
    this.match = validated ? validated[1] : '0000000000';
  }
  number() {
    return this.match;
  }
  areaCode() {
    return this.match.substr(0, 3);
  }
  toString() {
    const exchg = this.match.substr(3, 3);
    const num = this.match.substr(6, 4);
    return `(${this.areaCode()}) ${exchg}-${num}`;
  }

}

export default PhoneNumber;
